### PR TITLE
Pro 5763 fix apos mode corrupted

### PR DESCRIPTION
**Pull request by ValJed**
[PRO-5763](https://linear.app/apostrophecms/issue/PRO-5763/fix-corrupted-aposmode-when-inserting)

## Summary

* Fixes the docs `beforeInsert` handler (see ticket)
* Adds tests
* Adds migration to fix potential corrupted data.

## What are the specific steps to test this change?
* Tests must be green
* When inserting doc without ID or `aposMode` it should set mode based on `req.mode`.
* When inserting doc with an `_id`, `aposMode` and `aposLocale` are set based on the `_id`.
* When inserting doc without ID but with `aposLocale`, `aposMode` is set based on `aposLocale`.

## What kind of change does this PR introduce?

- [X] Bug fix
- [ ] New feature
- [ ] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other

## Make sure the PR fulfills these requirements:

- [X] It includes a) the existing issue ID being resolved, b) a convincing reason for adding this feature, or c) a clear description of the bug it resolves
- [X] The changelog is updated
- [ ] Related documentation has been updated
- [X] Related tests have been updated

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,12 @@
 # Changelog
 
+## UNRELEASED
+
+### Fixes
+
+* Updates the docs `beforeInsert` handler to avoid ending with different modes being set between `_id`, `aposLocale` and `aposMode`.
+* Adds a migration to fix potential corrupted data have different modes set between `_id`, `aposLocale` and `aposMode`.
```
Comment (boutell) [inline CHANGELOG.md:8]: have -> having
have different -> with different

File: modules/@apostrophecms/doc/index.js [hunk 2]
```diff
@@ -120,10 +119,23 @@ module.exports = {
...
+            doc.aposMode = mode;
+            return;
           }
+
+          const [ locale, mode ] = doc.aposLocale
+            ? doc.aposLocale.split(':')
+            : [ req.locale, req.mode ];
+
+          doc.aposLocale = `${locale}:${mode}`;
+          doc.aposMode = mode;
```
Comment (boutell) [inline modules/@apostrophecms/doc/index.js:138]: 😭 so good

File: modules/@apostrophecms/doc/lib/migrations.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+          if (idMode !== localeMode || idMode !== aposMode) {
+            try {
+              await self.apos.doc.db.updateOne({ _id: doc._id }, {
+                $set: {
+                  aposLocale: `${locale}:${idMode}`,
+                  aposMode: idMode
+                }
+              });
+            } catch (err) {
+              console.error('Error while running the set-document-modes migration: ', err);
```
Comment (boutell) [inline modules/@apostrophecms/doc/lib/migrations.js:46]: Let it throwwwww, let it throwwwwww

**Review by boutell — CHANGES_REQUESTED** [review]
Great stuff, tiny thing about error handling, let it throw

File: CHANGELOG.md (hunk 1 above)
Comment (ValJed) [reply CHANGELOG.md:8]: changes `have different` but did not find the first `have`

File: modules/@apostrophecms/doc/index.js (hunk 2 above)
Comment (ValJed) [reply modules/@apostrophecms/doc/index.js:138]: 😁

File: modules/@apostrophecms/doc/lib/migrations.js (hunk 3 above)
Comment (ValJed) [reply modules/@apostrophecms/doc/lib/migrations.js]: Try catch has been removed.

**Review by boutell — APPROVED** [review]

**Pull request merged**